File: src/odc_api_sandbox/workflows.py
from __future__ import annotations

import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path
from typing import Any

import httpx

from .client import BUILD_TERMINAL_STATUSES, OPERATION_TERMINAL_STATUSES, OdcClient
from .errors import OdcApiError
from .resolve import resolve_asset_key, resolve_environment_key
from .utils import _PRINT_LOCK, compact_dict, print_json, require_key


def wait_for(
    label: str,
    fetch: Any,
    terminal_statuses: set[str],
    *,
    interval_seconds: float,
    timeout_seconds: float,
) -> dict[str, Any]:
    deadline = time.monotonic() + timeout_seconds
    last_status: str | None = None

    while True:
        details = fetch()
        status = details.get("status")
        if status != last_status:
            with _PRINT_LOCK:
                print(f"{label}: {status}")
            last_status = status
        if status in terminal_statuses:
            return details
        if time.monotonic() >= deadline:
            raise OdcApiError(f"Timed out waiting for {label}; last response: {details}")
        time.sleep(interval_seconds)


def print_preflight_summary(asset: dict[str, Any], environment: dict[str, Any], revision: int) -> None:
    print_json(
        {
            "preflight": {
                "asset": compact_dict(
                    asset,
                    [
                        "assetKey",
                        "name",
                        "assetType",
                        "revision",
                        "tag",
                        "portfolioKey",
                        "createdAt",
                        "createdBy",
                    ],
                ),
                "environment": compact_dict(
                    environment,
                    [
                        "key",
                        "name",
                        "purpose",
                        "defaultDomain",
                        "region",
                        "hosting",
                        "status",
                        "portfolioKey",
                    ],
                ),
                "selectedRevision": revision,
            }
        }
    )


def preflight(client: OdcClient, asset_key: str, environment_key: str, revision: int | None) -> int:
    asset = client.get_asset(asset_key)
    environment = client.get_environment(environment_key)
    resolved_revision = revision if revision is not None else asset.get("revision")
    if not isinstance(resolved_revision, int):
        resolved_revision = client.latest_revision(asset_key)
    print_preflight_summary(asset, environment, resolved_revision)
    return resolved_revision


def print_dependency_summary(client: OdcClient, asset_key: str, revision: int, environment_key: str) -> None:
    graph = client.producer_graph(asset_key, revision, environment_key=environment_key)
    producers = graph.get("results") or []
    with _PRINT_LOCK:
        print(f"Dependencies ({len(producers)}):")
        for producer in producers:
            name = producer.get("name", "Unknown")
            producer_type = producer.get("type", "Unknown")
            status = producer.get("status", "Unknown")
            print(f"  - {name} ({producer_type}) - {status}")


def run_all_for_asset(
    client: OdcClient,
    asset_key: str,
    environment_key: str,
    revision: int | None,
    build_type: str,
    poll_interval: float,
    timeout_seconds: float,
) -> dict[str, Any]:
    resolved_key = resolve_asset_key(client, asset_key)
    resolved_environment_key = resolve_environment_key(client, environment_key)
    resolved_revision = preflight(client, resolved_key, resolved_environment_key, revision)
    print_dependency_summary(client, resolved_key, resolved_revision, resolved_environment_key)

    label_prefix = f"[{asset_key}] "

    build_response = client.start_build(resolved_key, resolved_revision, build_type)
    print_json({"build_started": build_response})
    build_key = require_key(build_response.get("buildKey"), "buildKey")
    build_details = wait_for(
        f"{label_prefix}build {build_key}",
        lambda: client.get_build(build_key),
        BUILD_TERMINAL_STATUSES,
        interval_seconds=poll_interval,
        timeout_seconds=timeout_seconds,
    )
    if build_details.get("status") != "Finished":
        raise OdcApiError(f"Build did not finish successfully: {build_details.get('status')}")

    deploy_response = client.deploy(resolved_key, resolved_revision, build_key, resolved_environment_key)
    print_json({"deploy_started": deploy_response})
    deploy_key = require_key(deploy_response.get("key"), "deployment operation key")
    deploy_details = wait_for(
        f"{label_prefix}deployment {deploy_key}",
        lambda: client.get_deployment(deploy_key),
        OPERATION_TERMINAL_STATUSES,
        interval_seconds=poll_interval,
        timeout_seconds=timeout_seconds,
    )
    if deploy_details.get("status") != "Finished":
        raise OdcApiError(f"Deployment did not finish successfully: {deploy_details.get('status')}")

    result = {
        "assetKey": resolved_key,
        "environmentKey": resolved_environment_key,
        "revision": resolved_revision,
        "build": build_details,
        "deployment": deploy_details,
    }
    print_json(result)
    return result


def _undeploy_one(
    client: OdcClient,
    asset_key: str,
    asset_name: str,
    environment_key: str,
    poll_interval: float,
    timeout_seconds: float,
) -> dict[str, Any]:
    with _PRINT_LOCK:
        print(f"\n=== Undeploying '{asset_name}' ===")
    try:
        response = client.undeploy(asset_key, environment_key)
        operation_key = require_key(response.get("key"), "deployment operation key")
        details = wait_for(
            f"[{asset_name}] undeploy {operation_key}",
            lambda: client.get_deployment(operation_key),
            OPERATION_TERMINAL_STATUSES,
            interval_seconds=poll_interval,
            timeout_seconds=timeout_seconds,
        )
        if details.get("status") != "Finished":
            raise OdcApiError(f"Undeploy did not finish successfully: {details.get('status')}")
        return {"app": asset_name, "assetKey": asset_key, "status": "success", "result": details}
    except (OdcApiError, httpx.HTTPError) as exc:
        with _PRINT_LOCK:
            print(f"error: [{asset_name}] {exc}", file=sys.stderr)
        return {"app": asset_name, "assetKey": asset_key, "status": "failed", "error": str(exc)}


def undeploy_all_in_environment(
    client: OdcClient,
    environment_key: str,
    poll_interval: float,
    timeout_seconds: float,
    max_parallel: int,
) -> list[dict[str, Any]]:
    resolved_environment_key = resolve_environment_key(client, environment_key)
    deployed_assets = client.list_deployed_assets(resolved_environment_key)
    if not deployed_assets:
        print(f"No deployed apps found in environment {resolved_environment_key}.")
        return []

    max_parallel = max(1, max_parallel)
    client.token()

    apps = []
    for asset in deployed_assets:
        asset_key = asset.get("key")
        if not asset_key:
            continue
        deployments = asset.get("deployments") or []
        name = next((d.get("name") for d in deployments if d.get("name")), None) or asset_key
        apps.append((asset_key, name))

    summary: list[dict[str, Any]] = []
    with ThreadPoolExecutor(max_workers=max_parallel) as executor:
        futures = {
            executor.submit(
                _undeploy_one,
                client,
                asset_key,
                asset_name,
                resolved_environment_key,
                poll_interval,
                timeout_seconds,
            ): asset_key
            for asset_key, asset_name in apps
        }
        for future in as_completed(futures):
            summary.append(future.result())

    return summary


def build_dependency_plan(
    client: OdcClient, asset_keys: list[str], environment_key: str
) -> list[list[tuple[str, int | None]]]:
    """Expand asset_keys with their producer dependencies into ordered deploy levels.

    Each level is a list of (asset_key, revision) pairs that can be deployed in
    parallel; every level must finish before the next one starts. A dependency
    shared by multiple apps is only included once, at the earliest level all of
    its consumers need it deployed by.
    """
    revisions: dict[str, int | None] = {}
    producers_of: dict[str, set[str]] = {}

    def record(node_key: str, revision: int | None) -> None:
        if node_key not in producers_of:
            producers_of[node_key] = set()
            revisions[node_key] = revision
        elif revision is not None and revisions.get(node_key) is None:
            revisions[node_key] = revision

    def visit_producer_node(node: dict[str, Any]) -> str:
        node_key = require_key(node.get("key"), "producer key")
        record(node_key, node.get("revision"))
        for child in node.get("producers") or []:
            child_key = visit_producer_node(child)
            producers_of[node_key].add(child_key)
        return node_key

    for asset_key in asset_keys:
        resolved_key = resolve_asset_key(client, asset_key)
        record(resolved_key, None)
        revision = revisions[resolved_key] or client.latest_revision(resolved_key)
        graph = client.producer_graph(
            resolved_key, revision, environment_key=environment_key, producer_type_filter="All"
        )
        for producer in graph.get("results") or []:
            child_key = visit_producer_node(producer)
            producers_of[resolved_key].add(child_key)

    levels: dict[str, int] = {}

    def level_of(node_key: str) -> int:
        if node_key in levels:
            return levels[node_key]
        levels[node_key] = 0  # guard against cycles
        deps = producers_of.get(node_key) or set()
        level = 1 + max((level_of(dep) for dep in deps), default=-1)
        levels[node_key] = level
        return level

    for node_key in producers_of:
        level_of(node_key)

    max_level = max(levels.values(), default=-1)
    plan: list[list[tuple[str, int | None]]] = [[] for _ in range(max_level + 1)]
    for node_key, level in levels.items():
        plan[level].append((node_key, revisions.get(node_key)))
    return plan


def read_apps_file(path: str) -> list[str]:
    apps_path = Path(path)
    if not apps_path.is_file():
        raise OdcApiError(f"Apps file not found: {path}")
    lines = apps_path.read_text(encoding="utf-8").splitlines()
    apps = [line.strip() for line in lines]
    apps = [app for app in apps if app and not app.startswith("#")]
    if not apps:
        raise OdcApiError(f"Apps file is empty: {path}")
    return apps


def _deploy_one(
    client: OdcClient,
    asset_key: str,
    environment_key: str,
    revision: int | None,
    build_type: str,
    poll_interval: float,
    timeout_seconds: float,
) -> dict[str, Any]:
    with _PRINT_LOCK:
        print(f"\n=== Deploying '{asset_key}' ===")
    try:
        result = run_all_for_asset(
            client,
            asset_key,
            environment_key,
            revision,
            build_type,
            poll_interval,
            timeout_seconds,
        )
        return {"app": asset_key, "status": "success", "result": result}
    except (OdcApiError, httpx.HTTPError) as exc:
        with _PRINT_LOCK:
            print(f"error: [{asset_key}] {exc}", file=sys.stderr)
        return {"app": asset_key, "status": "failed", "error": str(exc)}


def batch_deploy(
    client: OdcClient,
    apps_file: str,
    environment_key: str,
    revision: int | None,
    build_type: str,
    poll_interval: float,
    timeout_seconds: float,
    max_parallel: int,
    continue_on_error: bool,
    skip_dependencies: bool,
) -> list[dict[str, Any]]:
    apps = read_apps_file(apps_file)
    resolved_environment_key = resolve_environment_key(client, environment_key)
    max_parallel = max(1, max_parallel)

    # Force token acquisition once up front so concurrent workers don't race on it.
    client.token()

    if skip_dependencies:
        plan = [[(asset_key, revision) for asset_key in apps]]
    else:
        explicit_keys = {resolve_asset_key(client, asset_key) for asset_key in apps}
        plan = [
            # Only apply the requested revision to apps explicitly listed in the file;
            # dependencies deploy at the revision resolved from the producer graph.
            [(key, revision if key in explicit_keys else dep_revision) for key, dep_revision in level]
            for level in build_dependency_plan(client, apps, resolved_environment_key)
        ]
        added = sum(1 for level in plan for key, _ in level if key not in explicit_keys)
        if added:
            print(f"Including {added} dependency app(s) not listed in {apps_file}.")

    summary: list[dict[str, Any]] = []
    stop = False
    for level in plan:
        if stop:
            break
        if max_parallel == 1 and not continue_on_error:
            for asset_key, level_revision in level:
                entry = _deploy_one(
                    client,
                    asset_key,
                    resolved_environment_key,
                    level_revision,
                    build_type,
                    poll_interval,
                    timeout_seconds,
                )
                summary.append(entry)
                if entry["status"] == "failed":
                    stop = True
                    break
        else:
            with ThreadPoolExecutor(max_workers=max_parallel) as executor:
                futures = {
                    executor.submit(
                        _deploy_one,
                        client,
                        asset_key,
                        resolved_environment_key,
                        level_revision,
                        build_type,
                        poll_interval,
                        timeout_seconds,
                    ): asset_key
                    for asset_key, level_revision in level
                }
                for future in as_completed(futures):
                    summary.append(future.result())
            if not continue_on_error and any(entry["status"] == "failed" for entry in summary):
                stop = True

    return summary

File: internal/odc/workflows.go
package odc

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

type Options struct {
	JSON                                                           bool
	Color                                                          string
	App, Env, BuildType, BuildKey, Filter, Output, Search, AppType string
	Revision                                                       *int
	Interval, Timeout                                              time.Duration
	MaxParallel, MaxDepth                                          int
	NoWait, ContinueOnError, SkipDependencies, AllProducers        bool
	Updates                                                        object
}

func printlnLocked(format string, args ...any) {
	printMu.Lock()
	defer printMu.Unlock()
	w := os.Stdout
	if outputJSON {
		w = os.Stderr
	}
	fmt.Fprintf(w, format+"\n", args...)
}
func WaitFor(label string, fetch func() (object, error), build bool, interval, timeout time.Duration) (object, error) {
	deadline := time.Now().Add(timeout)
	last := ""
	for {
		details, e := fetch()
		if e != nil {
			return nil, e
		}
		status := str(details["status"])
		if status != last {
			printlnLocked("%s: %s", label, status)
			last = status
		}
		terminal := status == "Finished" || (!build && status == "FinishedWithError") || (build && (status == "FinishedWithErrors" || status == "Deleted" || status == "ToBeDeleted"))
		if terminal {
			return details, nil
		}
		remaining := time.Until(deadline)
		if remaining <= 0 {
			return nil, errorf("Timed out waiting for %s; last response: %v", label, details)
		}
		time.Sleep(min(interval, remaining))
	}
}
func (c *Client) Preflight(key, env string, revision *int) (int, error) {
	app, e := c.GetApp(key)
	if e != nil {
		return 0, e
	}
	environment, e := c.GetEnvironment(env)
	if e != nil {
		return 0, e
	}
	n, ok := integer(app["revision"])
	if revision != nil {
		n = *revision
	} else if !ok {
		n, e = c.LatestRevision(key)
		if e != nil {
			return 0, e
		}
	}
	e = PrintResult(object{"preflight": object{
		"app":         compactMap(app, []string{"assetKey", "name", "assetType", "revision", "tag", "portfolioKey", "createdAt", "createdBy"}),
		"environment": compactMap(environment, []string{"key", "name", "purpose", "defaultDomain", "region", "hosting", "status", "portfolioKey"}), "selectedRevision": n}})
	return n, e
}
func (c *Client) waitOperation(response object, kind string, o Options) (object, error) {
	keyField, api, path := "key", "deployments", "/deployment-operations/"
	if kind == "build" {
		keyField, api, path = "buildKey", "builds", "/build-operations/"
	} else if kind == "publish" {
		path = "/publish-operations/"
	}
	key, e := requireString(response[keyField], kind+" operation key")
	if e != nil {
		return nil, e
	}
	if o.NoWait {
		return response, nil
	}
	d, e := WaitFor(kind+" "+key, func() (object, error) { return c.call("GET", api, path+esc(key), nil, nil) }, kind == "build", o.Interval, o.Timeout)
	if e != nil {
		return nil, e
	}
	if d["status"] != "Finished" {
		return nil, errorf("%s did not finish successfully: %v", kind, d["status"])
	}
	return d, nil
}
func (c *Client) DeployApp(input, env string, revision *int, o Options) (object, error) {
	key, e := c.Resolve(input, "app")
	if e != nil {
		return nil, e
	}
	env, e = c.Resolve(env, "environment")
	if e != nil {
		return nil, e
	}
	rev, e := c.Preflight(key, env, revision)
	if e != nil {
		return nil, e
	}
	graph, e := c.ProducerGraph(key, rev, env, 0, "Deployable")
	if e != nil {
		return nil, e
	}
	producers := []object{}
	for _, p := range objects(graph["results"]) {
		pk, e := requireString(p["key"], "producer key")
		if e != nil {
			return nil, e
		}
		platform, e := c.IsPlatformProvided(pk)
		if e != nil {
			return nil, e
		}
		if !platform {
			producers = append(producers, p)
		}
	}
	printlnLocked("Dependencies (%d):", len(producers))
	for _, p := range producers {
		printlnLocked("  - %v (%v) - %v", first(p["name"], "Unknown"), first(p["type"], "Unknown"), first(p["status"], "Unknown"))
	}
	b, e := c.StartBuild(key, rev, o.BuildType)
	if e != nil {
		return nil, e
	}
	if e = PrintResult(object{"build_started": b}); e != nil {
		return nil, e
	}
	buildKey, e := requireString(b["buildKey"], "buildKey")
	if e != nil {
		return nil, e
	}
	build, e := c.waitOperation(b, "build", o)
	if e != nil {
		return nil, e
	}
	d, e := c.Deploy(key, rev, buildKey, env)
	if e != nil {
		return nil, e
	}
	if e = PrintResult(object{"deploy_started": d}); e != nil {
		return nil, e
	}
	deployment, e := c.waitOperation(d, "deployment", o)
	if e != nil {
		return nil, e
	}
	result := object{"assetKey": key, "environmentKey": env, "revision": rev, "build": build, "deployment": deployment}
	return result, PrintResult(result)
}

type App struct {
	Key      string
	Revision *int
}

func ReadAppsFile(path string) ([]App, error) {
	f, e := os.Open(path)
	if e != nil {
		return nil, errorf("Apps file: %v", e)
	}
	defer f.Close()
	apps := []App{}
	scanner := bufio.NewScanner(f)
	lineNumber := 0
	for scanner.Scan() {
		lineNumber++
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		key, rev, pinned := strings.Cut(line, "@")
		key = strings.TrimSpace(key)
		if key == "" {
			return nil, errorf("Empty app at %s:%d", path, lineNumber)
		}
		app := App{Key: key}
		if pinned {
			n, e := strconv.Atoi(strings.TrimSpace(rev))
			if e != nil || n < 1 {
				return nil, errorf("Invalid revision %q for app %q in %s", rev, key, path)
			}
			app.Revision = &n
		}
		apps = append(apps, app)
	}
	if e = scanner.Err(); e != nil {
		return nil, e
	}
	if len(apps) == 0 {
		return nil, errorf("Apps file is empty: %s", path)
	}
	return apps, nil
}
func (c *Client) DependencyPlan(apps []App, env string) ([][]App, error) {
	revisions := map[string]*int{}
	deps := map[string]map[string]bool{}
	record := func(key string, rev *int) error {
		if deps[key] == nil {
			deps[key] = map[string]bool{}
		}
		if old := revisions[key]; old != nil && rev != nil && *old != *rev {
			return errorf("Conflicting revisions for app %s: %d and %d", key, *old, *rev)
		}
		if rev != nil {
			revisions[key] = rev
		}
		return nil
	}
	resolved := []App{}
	for _, app := range apps {
		key, e := c.Resolve(app.Key, "app")
		if e != nil {
			return nil, e
		}
		rev := app.Revision
		if rev == nil {
			n, e := c.LatestRevision(key)
			if e != nil {
				return nil, e
			}
			rev = &n
		}
		if e = record(key, rev); e != nil {
			return nil, e
		}
		resolved = append(resolved, App{key, rev})
	}
	var visit func(object) (string, error)
	visit = func(node object) (string, error) {
		key, e := requireString(node["key"], "producer key")
		if e != nil {
			return "", e
		}
		n, ok := integer(node["revision"])
		if !ok || n < 1 {
			return "", errorf("Invalid producer revision for %s", key)
		}
		if e = record(key, &n); e != nil {
			return "", e
		}
		for _, child := range objects(node["producers"]) {
			ck, e := requireString(child["key"], "producer key")
			if e != nil {
				return "", e
			}
			platform, e := c.IsPlatformProvided(ck)
			if e != nil {
				return "", e
			}
			if platform {
				continue
			}
			ck, e = visit(child)
			if e != nil {
				return "", e
			}
			deps[key][ck] = true
		}
		return key, nil
	}
	for _, app := range resolved {
		g, e := c.ProducerGraph(app.Key, *app.Revision, env, 0, "All")
		if e != nil {
			return nil, e
		}
		for _, p := range objects(g["results"]) {
			key, e := requireString(p["key"], "producer key")
			if e != nil {
				return nil, e
			}
			platform, e := c.IsPlatformProvided(key)
			if e != nil {
				return nil, e
			}
			if platform {
				continue
			}
			key, e = visit(p)
			if e != nil {
				return nil, e
			}
			deps[app.Key][key] = true
		}
	}
	return dependencyLevels(revisions, deps)
}
func dependencyLevels(revisions map[string]*int, deps map[string]map[string]bool) ([][]App, error) {
	levels := map[string]int{}
	visiting := map[string]bool{}
	var levelOf func(string) (int, error)
	levelOf = func(key string) (int, error) {
		if n, ok := levels[key]; ok {
			return n, nil
		}
		if visiting[key] {
			return 0, errorf("Dependency cycle detected at app %s", key)
		}
		visiting[key] = true
		level := 0
		for dep := range deps[key] {
			n, e := levelOf(dep)
			if e != nil {
				return 0, e
			}
			level = max(level, n+1)
		}
		visiting[key] = false
		levels[key] = level
		return level, nil
	}
	keys := []string{}
	for key := range deps {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	plan := [][]App{}
	for _, key := range keys {
		n, e := levelOf(key)
		if e != nil {
			return nil, e
		}
		for len(plan) <= n {
			plan = append(plan, []App{})
		}
		plan[n] = append(plan[n], App{key, revisions[key]})
	}
	return plan, nil
}
func runParallel(apps []App, o Options, run func(App) object) []object {
	summary := []object{}
	if o.MaxParallel <= 1 {
		for _, app := range apps {
			entry := run(app)
			summary = append(summary, entry)
			if entry["status"] == "failed" && !o.ContinueOnError {
				break
			}
		}
		return summary
	}
	jobs := make(chan App)
	results := make(chan object, len(apps))
	var wg sync.WaitGroup
	for i := 0; i < min(o.MaxParallel, len(apps)); i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for app := range jobs {
				results <- run(app)
			}
		}()
	}
	go func() {
		for _, app := range apps {
			jobs <- app
		}
		close(jobs)
		wg.Wait()
		close(results)
	}()
	for entry := range results {
		summary = append(summary, entry)
	}
	return summary
}
func resultEntry(app string, result object, e error) object {
	if e != nil {
		stderr("error: [%s] %v\n", app, e)
		return object{"app": app, "status": "failed", "error": e.Error()}
	}
	return object{"app": app, "status": "success", "result": result}
}
func (c *Client) BatchDeploy(path string, o Options) ([]object, error) {
	apps, e := ReadAppsFile(path)
	if e != nil {
		return nil, e
	}
	env, e := c.Resolve(o.Env, "environment")
	if e != nil {
		return nil, e
	}
	if _, e = c.Token(); e != nil {
		return nil, e
	}
	plan := [][]App{apps}
	if !o.SkipDependencies {
		plan, e = c.DependencyPlan(apps, env)
		if e != nil {
			return nil, e
		}
		explicit := map[string]bool{}
		for _, app := range apps {
			key, e := c.Resolve(app.Key, "app")
			if e != nil {
				return nil, e
			}
			explicit[key] = true
		}
		added := 0
		for _, level := range plan {
			for _, app := range level {
				if !explicit[app.Key] {
					added++
				}
			}
		}
		if added > 0 {
			printlnLocked("Including %d dependency app(s) not listed in %s.", added, path)
		}
	}
	summary := []object{}
	for _, level := range plan {
		summary = append(summary, runParallel(level, o, func(app App) object {
			printlnLocked("\n=== Deploying '%s' ===", app.Key)
			result, e := c.DeployApp(app.Key, env, app.Revision, o)
			return resultEntry(app.Key, result, e)
		})...)
		if !o.ContinueOnError && hasFailed(summary) {
			break
		}
	}
	return summary, nil
}
func (c *Client) BatchUndeploy(path string, o Options) ([]object, error) {
	apps, e := ReadAppsFile(path)
	if e != nil {
		return nil, e
	}
	env, e := c.Resolve(o.Env, "environment")
	if e != nil {
		return nil, e
	}
	if _, e = c.Token(); e != nil {
		return nil, e
	}
	plan := [][]App{apps}
	if !o.SkipDependencies {
		plan, e = c.DependencyPlan(apps, env)
		if e != nil {
			return nil, e
		}
		explicit := map[string]bool{}
		for _, app := range apps {
			key, e := c.Resolve(app.Key, "app")
			if e != nil {
				return nil, e
			}
			explicit[key] = true
		}
		added := 0
		for _, level := range plan {
			for _, app := range level {
				if !explicit[app.Key] {
					added++
				}
			}
		}
		if added > 0 {
			printlnLocked("Including %d dependency app(s) not listed in %s.", added, path)
		}
	}
	summary := []object{}
	for i := len(plan) - 1; i >= 0; i-- {
		level := plan[i]
		summary = append(summary, runParallel(level, o, func(app App) object {
			printlnLocked("\n=== Undeploying '%s' ===", app.Key)
			d, e := c.Undeploy(app.Key, env)
			if e == nil {
				d, e = c.waitOperation(d, "undeploy", o)
			}
			return resultEntry(app.Key, d, e)
		})...)
		if !o.ContinueOnError && hasFailed(summary) {
			break
		}
	}
	return summary, nil
}
func (c *Client) BatchDelete(path string, o Options) ([]object, error) {
	apps, e := ReadAppsFile(path)
	if e != nil {
		return nil, e
	}
	plan := [][]App{apps}
	if !o.SkipDependencies {
		env, e := c.Resolve(o.Env, "environment")
		if e != nil {
			return nil, e
		}
		plan, e = c.DependencyPlan(apps, env)
		if e != nil {
			return nil, e
		}
		explicit := map[string]bool{}
		for _, app := range apps {
			key, e := c.Resolve(app.Key, "app")
			if e != nil {
				return nil, e
			}
			explicit[key] = true
		}
		added := 0
		for _, level := range plan {
			for _, app := range level {
				if !explicit[app.Key] {
					added++
				}
			}
		}
		if added > 0 {
			printlnLocked("Including %d dependency app(s) not listed in %s.", added, path)
		}
	}
	if _, e = c.Token(); e != nil {
		return nil, e
	}
	summary := []object{}
	for i := len(plan) - 1; i >= 0; i-- {
		level := plan[i]
		summary = append(summary, runParallel(level, o, func(app App) object {
			printlnLocked("\n=== Deleting '%s' ===", app.Key)
			key, e := c.Resolve(app.Key, "app")
			if e == nil {
				_, e = c.call("DELETE", "asset-repository", "/assets/"+esc(key), nil, nil)
			}
			return resultEntry(app.Key, object{"status": "success"}, e)
		})...)
		if !o.ContinueOnError && hasFailed(summary) {
			break
		}
	}
	return summary, nil
}
func (c *Client) UndeployAll(o Options) ([]object, error) {
	env, e := c.Resolve(o.Env, "environment")
	if e != nil {
		return nil, e
	}
	deployed, e := c.ListDeployedApps(env)
	if e != nil {
		return nil, e
	}
	apps := []App{}
	names := map[string]string{}
	for _, app := range deployed {
		key := str(app["key"])
		if key == "" {
			continue
		}
		name := key
		for _, d := range objects(app["deployments"]) {
			if str(d["name"]) != "" {
				name = str(d["name"])
				break
			}
		}
		apps = append(apps, App{Key: key})
		names[key] = name
	}
	if len(apps) == 0 {
		printlnLocked("No deployed apps found in environment %s.", env)
		return []object{}, nil
	}
	if _, e = c.Token(); e != nil {
		return nil, e
	}
	return runParallel(apps, o, func(app App) object {
		printlnLocked("\n=== Undeploying '%s' ===", names[app.Key])
		d, e := c.Undeploy(app.Key, env)
		if e == nil {
			d, e = c.waitOperation(d, "undeploy", o)
		}
		entry := resultEntry(names[app.Key], d, e)
		entry["assetKey"] = app.Key
		return entry
	}), nil
}
